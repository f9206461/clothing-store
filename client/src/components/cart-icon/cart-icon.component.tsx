import React from 'react';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';

import { toggleCartHidden } from '../../redux/cart/cart.actions';
import { selectCartItemsCount } from '../../redux/cart/cart.selectors';

import { CartIconContainer, ShoppingIconContainer, ItemCountContainer } from './cart-icon.styles';

interface Props {
    toggleCartHidden: () => void;
    itemCount: number;
}

export const CartIcon = ({ toggleCartHidden, itemCount }: Props) => (
    <CartIconContainer onClick={toggleCartHidden}>
        <ShoppingIconContainer />
        <ItemCountContainer>{itemCount}</ItemCountContainer>
    </CartIconContainer>
)

// Take state from rootReducer, then pass to the CartIcon as props
const mapStateToProps = createStructuredSelector({
    // mapStateToProps has a shallow equality check (for primitive data types). 
    // Hence, even without using selector, it will not re-render.
    itemCount: selectCartItemsCount
});

const mapDispatchToProps = (dispatch: any) => ({
    toggleCartHidden: () => dispatch(toggleCartHidden())
});

export default connect(mapStateToProps, mapDispatchToProps)(CartIcon);