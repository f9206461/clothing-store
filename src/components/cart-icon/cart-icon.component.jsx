import React from 'react';
import { connect } from 'react-redux';

import { toggleCartHidden } from '../../redux/cart/cart.actions';

import { ReactComponent as ShoppingIcon } from '../../assets/shopping-bag.svg';

import './cart-icon.styles.scss';

const CartIcon = ({ toggleCartHidden, numberOfCartItems }) => (
    <div className="cart-icon" onClick={toggleCartHidden}>
        <ShoppingIcon className='shopping-icon'/>
        <span className="item-count">{numberOfCartItems}</span>
    </div>
)

// Take state from rootReducer, then pass to the CartIcon as props
const mapStateToProps = (state) => ({
    numberOfCartItems: state.cart.cartItems.reduce((accu, item) => accu + item.quantity, 0)
});

const mapDispatchToProps = (dispatch) => ({
    toggleCartHidden: () => dispatch(toggleCartHidden())
});

export default connect(mapStateToProps, mapDispatchToProps)(CartIcon);