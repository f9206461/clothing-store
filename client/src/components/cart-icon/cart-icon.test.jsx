import React from "react";
import { shallow } from 'enzyme';
import { CartIcon } from "./cart-icon.component";

describe('cart-icon', () => {
    let wrapper;
    const mockToggleCartHidden = jest.fn();
    const mockItemCount = 15;

    beforeEach(() => {
        const mockProps = {
            toggleCartHidden: mockToggleCartHidden, 
            itemCount: mockItemCount
        }
        wrapper = shallow(<CartIcon { ...mockProps } />);
    });

    it ('should render CartIcon correctly', () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('should call toggleCartHidden when icon is clicked', () => {
        wrapper.find('CartIconContainer').simulate('click');
        expect(mockToggleCartHidden).toHaveBeenCalled();
    });
    
    it('should render the itemCount as the text', () => {
        const itemCount = parseInt(wrapper.find('ItemCountContainer').text());
        expect(itemCount).toBe(mockItemCount);
    });
});