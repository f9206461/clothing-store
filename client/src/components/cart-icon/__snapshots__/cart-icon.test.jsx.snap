// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cart-icon should render CartIcon correctly 1`] = `
<CartIconContainer
  onClick={[MockFunction]}
>
  <ShoppingIconContainer />
  <ItemCountContainer>
    15
  </ItemCountContainer>
</CartIconContainer>
`;
